tensorflow/lite/kernels/stablehlo_reduce_precision.cc

```diff
@@ -0,0 +1,195 @@
+                                        31) +
+                                       1;
+  using UIntType =
+      std::conditional_t<std::is_same_v<DataType, float>, uint32_t, uint16_t>;
+};
+
+template <typename UIntType>
+UIntType ReducePrecision(UIntType int_val, int32_t exponent_bits,
+                         int32_t mantissa_bits, int32_t src_mantissa_bits,
+                         int32_t src_exponent_bits,
+                         UIntType last_mantissa_bit_mask,
+                         UIntType base_rounding_bias, UIntType truncation_mask,
+                         UIntType sign_bit_mask, UIntType exp_bits_mask,
+                         UIntType reduced_max_exponent,
+                         UIntType reduced_min_exponent);
+
+template <typename DataType>
+TfLiteStatus ReducePrecisionOp(const TfLiteTensor* operand,
+                               int32_t exponent_bits, int32_t mantissa_bits,
+                               TfLiteTensor* result) {
+  const int num_elements = NumElements(result);
+  const DataType* input = GetTensorData<DataType>(operand);
+  DataType* output = GetTensorData<DataType>(result);
+  using UIntType = typename FloatingPointTraits<DataType>::UIntType;
+  const int32_t src_mantissa_bits =
+      FloatingPointTraits<DataType>::kMantissaBits;
+  const int32_t src_exponent_bits =
+      FloatingPointTraits<DataType>::kExponentBits;
+
+  UIntType last_mantissa_bit_mask = static_cast<UIntType>(1)
+                                    << (src_mantissa_bits - mantissa_bits);
+  UIntType base_rounding_bias = (last_mantissa_bit_mask >> 1) - 1;
+  UIntType truncation_mask = ~(last_mantissa_bit_mask - 1);
+  UIntType sign_bit_mask = static_cast<UIntType>(1)
+                           << (sizeof(UIntType) * 8 - 1);
+  UIntType exp_bits_mask = ((static_cast<UIntType>(1) << src_exponent_bits) - 1)
+                           << src_mantissa_bits;
+  UIntType exponent_bias =
+      (static_cast<UIntType>(1) << (src_exponent_bits - 1)) - 1;
+  UIntType reduced_exponent_bias =
+      (static_cast<UIntType>(1) << (exponent_bits - 1)) - 1;
+  UIntType reduced_max_exponent = exponent_bias + reduced_exponent_bias;
+  UIntType reduced_min_exponent = exponent_bias - reduced_exponent_bias;
+  for (int i = 0; i < num_elements; ++i) {
+    UIntType int_val;
+    std::memcpy(&int_val, &input[i], sizeof(UIntType));
+    int_val = ReducePrecision(int_val, exponent_bits, mantissa_bits,
+                              src_mantissa_bits, src_exponent_bits,
+                              last_mantissa_bit_mask, base_rounding_bias,
+                              truncation_mask, sign_bit_mask, exp_bits_mask,
+                              reduced_max_exponent, reduced_min_exponent);
+    std::memcpy(&output[i], &int_val, sizeof(DataType));
+    if (std::isnan(input[i])) {
+      output[i] = mantissa_bits > 0 ? input[i]
+                                    : std::numeric_limits<DataType>::infinity();
+    }
```

Comment:
This indicates that the ReducePrecision call is unnecessary if the input is NaN. Can we move this check before and elide the ReducePrecision call?

Reply:
yeah sure